Fix c-ares crash

arthurpassos opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

A segmentation fault related to DNS & c-ares has been reported. The below error ocurred in multiple threads:
```
2022-09-28 15:41:19.008,2022.09.28 15:41:19.008088 [ 356 ] {} <Fatal> BaseDaemon: ########################################
2022-09-28 15:41:19.008,"2022.09.28 15:41:19.008147 [ 356 ] {} <Fatal> BaseDaemon: (version 22.8.5.29 (official build), build id: 92504ACA0B8E2267) (from thread 353) (no query) Received signal Segmentation fault (11)"
2022-09-28 15:41:19.008,2022.09.28 15:41:19.008196 [ 356 ] {} <Fatal> BaseDaemon: Address: 0xf Access: write. Address not mapped to object.
2022-09-28 15:41:19.008,2022.09.28 15:41:19.008216 [ 356 ] {} <Fatal> BaseDaemon: Stack trace: 0x188f8212 0x1626851b 0x1626a69e 0x16269b3f 0x16267eab 0x13cf8284 0x13d24afc 0x13c5217e 0x14ec2495 0x15ba440f 0x15b9d13b 0x15bb2699 0x1891ccb3 0x1891e00d 0x18ae0769 0x18ade022 0x7f76aa985609 0x7f76aa8aa133
2022-09-28 15:41:19.008,2022.09.28 15:41:19.008274 [ 356 ] {} <Fatal> BaseDaemon: 2. Poco::Net::IPAddress::family() const @ 0x188f8212 in /usr/bin/clickhouse
2022-09-28 15:41:19.008,2022.09.28 15:41:19.008297 [ 356 ] {} <Fatal> BaseDaemon: 3. ? @ 0x1626851b in /usr/bin/clickhouse
2022-09-28 15:41:19.008,2022.09.28 15:41:19.008309 [ 356 ] {} <Fatal> BaseDaemon: 4. ? @ 0x1626a69e in /usr/bin/clickhouse
```

In a proper scenario (DNS server and host_regexp config), an "auth storm" reproduces the issue. I used `clickhouse-benchmark` to perform it. Command: `echo 'select 1' | clickhouse benchmark -h <host> --user test_dns -c 85 --reconnect`.

I checked the docs for [ares_fds](https://c-ares.org/ares_fds.html) and [ares_process](https://c-ares.org/ares_process.html), I don't see any mention of it being non-thread safe. While the docs for [ares_library_init](https://c-ares.org/ares_library_init.html) explicitly mentions it's not thread safe.

There isn't much documentation about `c-ares` and its usage. The examples in the docs are pretty much equivalent to what we have implemented. I have spent some time looking into how other famous projects like `nodejs` and `wireshark` make use of `c-ares`. The former does [have some sort of mutex](https://github.com/nodejs/node/blob/eac0e59bc2689a16e6e8b53c85bb52a1e28329b4/src/cares_wrap.cc#L78) but seems to use it only for [initialization](https://github.com/nodejs/node/blob/eac0e59bc2689a16e6e8b53c85bb52a1e28329b4/src/cares_wrap.cc#L722). Furthermore, the implementation is somewhat abstracted and not so straight forward to understand. Wireshark, on the other hand, implements it in a very simple & similar fashion to what Clickhouse CaresPTRResolver is doing. There are some differences, tho. Still, I couldn't find a mutex usage or even multi-threading mentions.

Going through the issues in the github repo, I see issues like [ares_* not thread safe with respect to the same channel](https://github.com/c-ares/c-ares/issues/208) and [ares_library_init should be thread safe](https://github.com/c-ares/c-ares/issues/449). Both have mentions of c-ares not being thread safe and that should be handled by the client. I filed a github [issue](https://github.com/c-ares/c-ares/issues/484) asking for clarification, but still no response.

I looked through the implementation of the methods we call (`ares_fds`, `ares_timeout`, `select` and `ares_process`) and it does seem to touch any global / shared state. Meaning I don't understand why the race condition is happening.

I want to add a test that does the auth storm to make sure such thing doesn't happen again, but I am not sure how to do it. It has to be an integration test because it requires the whole DNS server setup. `clickhouse-benchmark` seems to be accessible, but it's hard to validate the output and when doing the auth storm other issues appear. One exception I observed while running a dummy test with ch-benchmark is: `DB::NetException: Net Exception: Cannot assign requested address: 10.0.0.2:9000`

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

arthurpassos committed:
Add mutex around ares fd processing

arthurpassos commented:
@tavplubix maybe you could help me on this one?

arthurpassos commented:
It's worth nothing the below `gtest` doesn't reproduce the issue, not sure why tho:

```
TEST(Common, ReverseDNS)
{
    auto func = []()
    {
        for (int i = 0; i < 900u; i++) {
            auto resolver = DNSPTRResolverProvider::get();
            auto result = resolver->resolve("172.16.0.2");
            assert(!result.empty());
        }
    };

    auto number_of_threads = 1000u;
    std::vector<std::thread> threads;
    threads.reserve(number_of_threads);

    for (auto i = 0u; i < number_of_threads; i++)
    {
        threads.emplace_back(func);
    }

    for (auto & thread : threads)
    {
        thread.join();
    }
}
```

arthurpassos committed:
add mutex include for release builds

arthurpassos commented:
I have just implemented the file descriptor wait & processing using `ares_getsock`, `poll` and `ares_process_fds` instead of `ares_fds`, `ares_timeout`, `select`and `ares_process` and it doesn't seem to crash anymore. Both benchmark and the above gtest will fail with `DB::NetException: Net Exception: Cannot assign requested address: 172.16.0.1:9000`. It seems like one of the DNS queries do fail, I am not sure if that's expected or if there is something wrong with my draft implementation. At a second thought, I am not even sure the current master branch implementation will pass through that unit tests. Will give it a try.

m1khal3v commented:
Looks like we have same problem on our test 22.8 cluster

```
ch2-2    | 2022.10.12 08:06:12.914936 [ 271 ] {} <Fatal> BaseDaemon: ########################################
ch2-2    | 2022.10.12 08:06:12.914980 [ 271 ] {} <Fatal> BaseDaemon: (version 22.8.6.71 (official build), build id: 68D8301D2C882600) (from thread 197) (no query) Received signal Segmentation fault (11)
ch2-2    | 2022.10.12 08:06:12.915020 [ 271 ] {} <Fatal> BaseDaemon: Address: 0x10 Access: read. Address not mapped to object.
ch2-2    | 2022.10.12 08:06:12.915031 [ 271 ] {} <Fatal> BaseDaemon: Stack trace: 0xa4f4259 0xa4f7b08 0xa4f6f5c 0xa4f598c 0x148e72da 0x13e9d3b8 0x13e9fc96 0x13ea09cc 0xa4ae1a8 0xa4b13bd 0x7f0904b1a609 0x7f0904a3f133
ch2-2    | 2022.10.12 08:06:12.915239 [ 271 ] {} <Fatal> BaseDaemon: 2. ? @ 0xa4f4259 in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915288 [ 271 ] {} <Fatal> BaseDaemon: 3. DB::DNSResolver::updateAddress(Poco::Net::IPAddress const&) @ 0xa4f7b08 in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915361 [ 271 ] {} <Fatal> BaseDaemon: 4. bool DB::DNSResolver::updateCacheImpl<bool (DB::DNSResolver::*)(Poco::Net::IPAddress const&), std::__1::unordered_map<Poco::Net::IPAddress, unsigned int, std::__1::hash<Poco::Net::IPAddress>, std::__1::equal_to<Poco::Net::IPAddress>, std::__1::allocator<std::__1::pair<Poco::Net::IPAddress const, unsigned int> > >&>(bool (DB::DNSResolver::*&&)(Poco::Net::IPAddress const&), std::__1::unordered_map<Poco::Net::IPAddress, unsigned int, std::__1::hash<Poco::Net::IPAddress>, std::__1::equal_to<Poco::Net::IPAddress>, std::__1::allocator<std::__1::pair<Poco::Net::IPAddress const, unsigned int> > >&, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xa4f6f5c in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915389 [ 271 ] {} <Fatal> BaseDaemon: 5. DB::DNSResolver::updateCache(unsigned int) @ 0xa4f598c in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915550 [ 271 ] {} <Fatal> BaseDaemon: 6. DB::DNSCacheUpdater::run() @ 0x148e72da in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915676 [ 271 ] {} <Fatal> BaseDaemon: 7. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x13e9d3b8 in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915695 [ 271 ] {} <Fatal> BaseDaemon: 8. DB::BackgroundSchedulePool::threadFunction() @ 0x13e9fc96 in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915705 [ 271 ] {} <Fatal> BaseDaemon: 9. ? @ 0x13ea09cc in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915774 [ 271 ] {} <Fatal> BaseDaemon: 10. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa4ae1a8 in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915791 [ 271 ] {} <Fatal> BaseDaemon: 11. ? @ 0xa4b13bd in /usr/bin/clickhouse
ch2-2    | 2022.10.12 08:06:12.915954 [ 271 ] {} <Fatal> BaseDaemon: 12. ? @ 0x7f0904b1a609 in ?
ch2-2    | 2022.10.12 08:06:12.916018 [ 271 ] {} <Fatal> BaseDaemon: 13. __clone @ 0x7f0904a3f133 in ?
```

arthurpassos commented:
> Looks like we have same problem on our test 22.8 cluster
> 
> ```
> ch2-2    | 2022.10.12 08:06:12.914936 [ 271 ] {} <Fatal> BaseDaemon: ########################################
> ch2-2    | 2022.10.12 08:06:12.914980 [ 271 ] {} <Fatal> BaseDaemon: (version 22.8.6.71 (official build), build id: 68D8301D2C882600) (from thread 197) (no query) Received signal Segmentation fault (11)
> ch2-2    | 2022.10.12 08:06:12.915020 [ 271 ] {} <Fatal> BaseDaemon: Address: 0x10 Access: read. Address not mapped to object.
> ch2-2    | 2022.10.12 08:06:12.915031 [ 271 ] {} <Fatal> BaseDaemon: Stack trace: 0xa4f4259 0xa4f7b08 0xa4f6f5c 0xa4f598c 0x148e72da 0x13e9d3b8 0x13e9fc96 0x13ea09cc 0xa4ae1a8 0xa4b13bd 0x7f0904b1a609 0x7f0904a3f133
> ch2-2    | 2022.10.12 08:06:12.915239 [ 271 ] {} <Fatal> BaseDaemon: 2. ? @ 0xa4f4259 in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915288 [ 271 ] {} <Fatal> BaseDaemon: 3. DB::DNSResolver::updateAddress(Poco::Net::IPAddress const&) @ 0xa4f7b08 in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915361 [ 271 ] {} <Fatal> BaseDaemon: 4. bool DB::DNSResolver::updateCacheImpl<bool (DB::DNSResolver::*)(Poco::Net::IPAddress const&), std::__1::unordered_map<Poco::Net::IPAddress, unsigned int, std::__1::hash<Poco::Net::IPAddress>, std::__1::equal_to<Poco::Net::IPAddress>, std::__1::allocator<std::__1::pair<Poco::Net::IPAddress const, unsigned int> > >&>(bool (DB::DNSResolver::*&&)(Poco::Net::IPAddress const&), std::__1::unordered_map<Poco::Net::IPAddress, unsigned int, std::__1::hash<Poco::Net::IPAddress>, std::__1::equal_to<Poco::Net::IPAddress>, std::__1::allocator<std::__1::pair<Poco::Net::IPAddress const, unsigned int> > >&, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xa4f6f5c in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915389 [ 271 ] {} <Fatal> BaseDaemon: 5. DB::DNSResolver::updateCache(unsigned int) @ 0xa4f598c in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915550 [ 271 ] {} <Fatal> BaseDaemon: 6. DB::DNSCacheUpdater::run() @ 0x148e72da in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915676 [ 271 ] {} <Fatal> BaseDaemon: 7. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x13e9d3b8 in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915695 [ 271 ] {} <Fatal> BaseDaemon: 8. DB::BackgroundSchedulePool::threadFunction() @ 0x13e9fc96 in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915705 [ 271 ] {} <Fatal> BaseDaemon: 9. ? @ 0x13ea09cc in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915774 [ 271 ] {} <Fatal> BaseDaemon: 10. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa4ae1a8 in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915791 [ 271 ] {} <Fatal> BaseDaemon: 11. ? @ 0xa4b13bd in /usr/bin/clickhouse
> ch2-2    | 2022.10.12 08:06:12.915954 [ 271 ] {} <Fatal> BaseDaemon: 12. ? @ 0x7f0904b1a609 in ?
> ch2-2    | 2022.10.12 08:06:12.916018 [ 271 ] {} <Fatal> BaseDaemon: 13. __clone @ 0x7f0904a3f133 in ?
> ```
> 
> and few lines before
> 
> ```
> ch2-2    | 2022.10.12 08:06:12.903886 [ 197 ] {} <Error> DNSResolver: Cannot resolve host (ch3-2), error 0: Host not found.
> ch2-2    | 2022.10.12 08:06:12.911997 [ 197 ] {} <Information> DNSResolver: Cached hosts not found: ch2-1, ch3-2
> ch2-2    | 2022.10.12 08:06:12.912026 [ 197 ] {} <Information> DNSResolver: Cached hosts dropped: ch2-1, ch3-2
> ```

Do you have DNS cache disabled?

m1khal3v commented:
> Do you have DNS cache disabled?

No, it's enabled, i think. How i can check this? 

nslookup clickhouse.com in container:
```
Server:		127.0.0.11
Address:	127.0.0.11#53
```

nslookup clickhouse.com on host:
```
Server:		127.0.0.53
Address:	127.0.0.53#53
```

disable_internal_dns_cache is not used in config and cant find it in system.settings
```
SELECT *
FROM system.settings
WHERE name = 'disable_internal_dns_cache'

Query id: 105241bb-1bc4-4d7b-854e-08d013ba9975

Ok.

0 rows in set. Elapsed: 0.003 sec. 
```

arthurpassos committed:
Use poll instead of select for c-ares

arthurpassos commented:
I have spent some more time comparing three approaches and here is the scenario breakdown alongside the conclusions so far.

Scenario:
DNS server configured in a very similar fashion to what can be found in `tests/integration/test_host_regexp_multiple_ptr_records`.

CH server running locally with:
```
<disable_internal_dns_cache>1</disable_internal_dns_cache>
<max_concurrent_queries>200</max_concurrent_queries>

<yandex>
    <users>
        <test_dns>
            <password/>
            <networks>
                <host_regexp>test3\.example\.com$</host_regexp>
                <host_regexp>aftermath\.example\.com$</host_regexp>
            </networks>
            <profile>default</profile>
        </test_dns>
    </users>
</yandex>
```
Test tools:
* clickhouse-benchmark, command executed: time echo 'select 1' | clickhouse benchmark -h 172.16.0.1 --user test_dns -c 85 --reconnect
* custom python script: 
```
import pycurl
import threading
from io import BytesIO

mutex = threading.Lock()
success_counter = 0

def perform_request():

    buffer = BytesIO()
    crl = pycurl.Curl()
    crl.setopt(pycurl.INTERFACE, '192.168.0.157')
    crl.setopt(crl.WRITEDATA, buffer)
    crl.setopt(crl.URL, 'http://127.0.0.1:8123/?query=select+1&user=test_dns')

    crl.perform()

    # End curl session
    crl.close()

    str_response = buffer.getvalue().decode('iso-8859-1')
    expected_response = "1\n"

    mutex.acquire()

    global success_counter

    if (str_response == expected_response):
        success_counter += 1

    mutex.release()

   # print(buffer.getvalue().decode('iso-8859-1'))

def perform_multiple_requests(n):
    for i in range(n):
        perform_request()

threads = []


for i in range(195):
    thread = threading.Thread(target=perform_multiple_requests, args=(400,))
    thread.start()
    threads.append(thread)


for thread in threads:
    thread.join()

print("Success counter: " + str(success_counter))
```

* unit-test(200 threads, 200 iterations each)
* unit-test(500 threads, 500 iterations each)
```
TEST(Common, ReverseDNS)
{
    auto func = []()
    {
        for (int i = 0; i < 200u; i++) {
            auto resolver = DNSPTRResolverProvider::get();
            auto result = resolver->resolve("192.168.0.157");
            assert(!result.empty());
        }
    };

    auto number_of_threads = 200u;
    std::vector<std::thread> threads;
    threads.reserve(number_of_threads);

    for (auto i = 0u; i < number_of_threads; i++)
    {
        threads.emplace_back(func);
    }

    for (auto & thread : threads)
    {
        thread.join();
    }
}
```

Note: The above unit-test doesn't crash even in the current master branch implementation (no race condition protection at all). The other two tests do reproduce the crash in current master branch impl.

Approaches:
* `select` with a mutex on the wait method, first commit of this PR.
* lock free poll implementation, current state of this PR.
* mutex poll implementation, the above poll implementation with a mutex at the beginning of `CaresPTRResolver::resolve(ip)`

**Lock free poll**
|                       | crash | resolution/auth failure                                                                           | benchmark                                    |
|-----------------------|-------|---------------------------------------------------------------------------------------------------|----------------------------------------------|
| custom python script  | no    | no                                                                                                | real 0m55,207s, user 0m7,303s, sys 0m26,002s |
| clickhouse-benchmark  | no    | yes - failed after 27965 queries executed (successfully?), error: Cannot assign requested address | real 0m17.401s, user 0m5.135s, sys 0m24.880s |
| unit-test(200t, 200i) | no    | no                                                                                                | 820ms                                        |
| unit-test(500t, 500t) | no    | yes - 24 resolution failures                                                                      |                                              |

**Mutex poll**
|                       | crash | resolution/auth failure                                                                           | benchmark                                       |
|-----------------------|-------|---------------------------------------------------------------------------------------------------|-------------------------------------------------|
| custom python script  | no    | no                                                                                                | real  0m54,737s, user 0m7,637s, sys   0m25,157s |
| clickhouse-benchmark  | no    | yes - failed after 27034 queries executed (successfully?), error: Cannot assign requested address | real0m19.394s, user 0m5.391s, sys 0m27.402s     |
| unit-test(200t, 200i) | no    | no                                                                                                | 4s                                              |
| unit-test(500t, 500t) | no    | no                                                                                                | 27s                                             |

**Mutex select**
|                       | crash | resolution/auth failure                                                                                                                                                                        | benchmark                                    |
|-----------------------|-------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------|
| custom python script  | yes   |                                                                                                                                                                                                |                                              |
| clickhouse-benchmark  | no    | failed after 27915 queries executed (successfully?), two errors observed: Cannot assign requested address AND Authentication failed: password is incorrect or there is no user with such name. | real 1m45.033s, user 0m4.926s, sys 0m25.839s |
| unit-test(200t, 200i) | no    | no                                                                                                                                                                                             | 2s                                           |
| unit-test(500t, 500t) | no    | no                                                                                                                                                                                             | 14s                                          |

`Mutex select` is the only implementation that crashes. Thus, it's out of the equation. Out of the lock free and non lock free poll implementation, the non lock free seems to be the best one. No crashes and no failures. Surprisingly, it even performed better when it comes to speed.


I will spend some more time on this to get to a better conclusion, any input would be appreciated.

arthurpassos committed:
remove mutex declaration

arthurpassos committed:
working on style

arthurpassos committed:
working on style

arthurpassos committed:
working on style

arthurpassos committed:
custom py script in integ tests

arthurpassos committed:
Merge branch 'master' into fix_ares_fd_process_race_condition

arthurpassos commented:
@tavplubix can you enable the CI run?

arthurpassos committed:
style fix?

arthurpassos committed:
fix black style

arthurpassos commented:
The integ tests I just added is failing and the logs don't tell me much. I still can't run it locally because it requires the `pycurl` dependency to be available in the clickhouse-test docker image. I added and re-rebuilt the image, but it's not being picked up by the runner.

@tavplubix @Felixoid Can any of you assist me on running it locally?

arthurpassos commented:
> > Can any of you assist me on running it locally?
> 
> I don't use `runner` locally and prefer to run integration tests using pytest directly like this:
> 
> ```
> ~/ch/ClickHouse/tests/integration$ python3 -m pytest test_lost_part/test.py --count 1 -x -k test_lost_last
> ```
> 
> It requires to create some symlinks like
> 
> ```
> ~/ch/ClickHouse/tests/integration$ ls -l /etc/clickhouse-server/
> total 0
> lrwxrwxrwx 1 root root 56 Mar 14  2022 config.xml -> /home/tavplubix/ch/ClickHouse/programs/server/config.xml
> lrwxrwxrwx 1 root root 55 Mar 14  2022 users.xml -> /home/tavplubix/ch/ClickHouse/programs/server/users.xml
> ~/ch/ClickHouse/tests/integration$ ls -l /usr/bin/clickhouse*
> lrwxrwxrwx 1 root root   67 Jul  6 15:46 /usr/bin/clickhouse -> /home/tavplubix/ch/ClickHouse/cmake-build-debug/programs/clickhouse
> lrwxrwxrwx 1 root root   19 May  9 14:22 /usr/bin/clickhouse-benchmark -> /usr/bin/clickhouse
> lrwxrwxrwx 1 root root   19 Mar  7  2022 /usr/bin/clickhouse-client -> /usr/bin/clickhouse
> lrwxrwxrwx 1 root root   19 May 24 18:51 /usr/bin/clickhouse-format -> /usr/bin/clickhouse
> lrwxrwxrwx 1 root root   67 Apr 28 20:18 /usr/bin/clickhouse-keeper -> /home/tavplubix/ch/ClickHouse/cmake-build-debug/programs/clickhouse
> lrwxrwxrwx 1 root root   19 Mar  7  2022 /usr/bin/clickhouse-local -> /usr/bin/clickhouse
> lrwxrwxrwx 1 root root   19 Mar  7  2022 /usr/bin/clickhouse-server -> /usr/bin/clickhouse
> lrwxrwxrwx 1 root root   51 Mar 25  2022 /usr/bin/clickhouse-test -> /home/tavplubix/ch/ClickHouse/tests/clickhouse-test
> ```
> 
> And (the main disadvantage of this method) it requires to install some python packages locally (like in https://github.com/ClickHouse/ClickHouse/blob/bcd9453f332743d5fa0afc937210e6ea01d522cf/docker/test/integration/runner/Dockerfile). But most of these packages are not actually required if you don't run specific tests which use them:
> 
> https://github.com/ClickHouse/ClickHouse/blob/4b4c63749afcca4fd8b111618690bb2a0619f581/tests/integration/helpers/cluster.py#L21-L39
> 
> 
> (you can temporarily or permanently move other packages to this list if you cannot install them for some reason)

It seems like not even that helped. Running `docker images`, I see my `clickhouse/integration-base` is from 4 weeks ago. If I delete it and rebuild it, I get a fresh image with the recently build dependencies. When I run `pytest`, dependency is not there. Running `docker images` again shows the image got overriden.

tavplubix commented:
> It seems like not even that helped. Running docker images, I see my clickhouse/integration-base is from 4 weeks ago. If I delete it and rebuild it, I get a fresh image with the recently build dependencies. When I run pytest, dependency is not there. Running docker images again shows the image got overriden.

You should install packages (including pycurl) locally if you run tests with pytest directly, `clickhouse/integration-base` is not required for this at all.

tavplubix commented:
Hm, yes, `pytest` runs `clickhouse-server`/`clickhouse-keeper` inside `clickhouse/integration-test`, but code of `pytest` and the test itself (`tests/integration/test_host_regexp_multiple_ptr_records_concurrent/test.py`) are executed locally without any containers when you run integration tests without `runner` (`runner` runs special container that runs `pytest` inside and then tests run containers with server/keeper, so with runner you get docker inside docker, but without `runner` there's only one layer and python code is executed outside). So this line
https://github.com/ClickHouse/ClickHouse/blob/f414edebae69c5d247f2657c5a312c2e06b13e98/tests/integration/test_host_regexp_multiple_ptr_records_concurrent/test.py#L1

is executed from outside of any containers and it looks for the package installed on your host system. 

> I have pycurl installed locally

Is there a chance it was installed for different python version or in different virtualenv or something? Is it imported successfully when you run something like `sudo python3 -c "import pycurl"`?

tavplubix commented:
> I don't follow it. How is it executed outside the containers? The environment described is based on different containers with different IPs.

Integration tests are just a bunch of python scripts. When you run something like
```
sudo python3 -m pytest test_host_regexp_multiple_ptr_records_concurrent/test.py -k test_host_regexp_multiple_ptr_records_concurrent
```
your machine simply runs these python scripts. When you do `cluster.start()` in `started_cluster` fixture, the test runs containers (see cluster.py) which you asked to run by calling `add_instance` and passing parameters like `with_coredns=True`. So all instances of `clickhouse-server` are run in containers, but all code of the test itself is executed in your host system just like ordinary python program. 

> This would explain why it's timing out, tho. My machine can't reach the docker containers.

Containers should be accessible, see `http_query_and_get_answer_with_error` (it sends http request from outside of containers to a server inside container) and client.py (it runs `clickhouse-client` from outside of containers and it successfully connects to a server inside container)

arthurpassos commented:
> > I don't follow it. How is it executed outside the containers? The environment described is based on different containers with different IPs.
> 
> Integration tests are just a bunch of python scripts. When you run something like
> 
> ```
> sudo python3 -m pytest test_host_regexp_multiple_ptr_records_concurrent/test.py -k test_host_regexp_multiple_ptr_records_concurrent
> ```
> 
> your machine simply runs these python scripts. When you do `cluster.start()` in `started_cluster` fixture, the test runs containers (see cluster.py) which you asked to run by calling `add_instance` and passing parameters like `with_coredns=True`. So all instances of `clickhouse-server` are run in containers, but all code of the test itself is executed in your host system just like ordinary python program.
> 
> > This would explain why it's timing out, tho. My machine can't reach the docker containers.
> 
> Containers should be accessible, see `http_query_and_get_answer_with_error` (it sends http request from outside of containers to a server inside container) and client.py (it runs `clickhouse-client` from outside of containers and it successfully connects to a server inside container)

Ok, maybe there is a confusion. I do understand that the `test.py` file is executed locally, but my if you look at the contents of `test.py`, it uploads a `stress_test.py` file to the client machine. That script is the one that depends on `pycurl`. Once it's uploaded, it executes it inside the container. Does that make sense?

tavplubix commented:
> Ok, maybe there is a confusion. I do understand that the test.py file is executed locally, but if you look at the contents of test.py, it uploads a stress_test.py file to the client machine. That script is the one that depends on pycurl. Once it's uploaded, it executes it inside the container. Does that make sense?

Oh, yes, sorry, I did not notice it.

Btw, which ip does `clickhouse-server` observe when request is sent from outside? Does it make sense to write that ip into coredns config, so it will be resolved to `test3.example.com` and it will be possible to send requests from outside?

arthurpassos commented:
> > Ok, maybe there is a confusion. I do understand that the test.py file is executed locally, but if you look at the contents of test.py, it uploads a stress_test.py file to the client machine. That script is the one that depends on pycurl. Once it's uploaded, it executes it inside the container. Does that make sense?
> 
> Oh, yes, sorry, I did not notice it.
> 
> Btw, which ip does `clickhouse-server` observe when request is sent from outside? Does it make sense to write that ip into coredns config, so it will be resolved to `test3.example.com` and it will be possible to send requests from outside?

I am not sure I understand, but I guess the main thing here is that this very same setup already works for other tests. Check `test_host_regexp_multiple_ptr_records`. The only problem should be the `pycurl` and possibly the other python dependencies of the `stress_test.py` script not being available in the machine.

I have tried several different workarounds to run this, but failed to do so. Couple of ones: build image locally, copy it to the docker container and run docker load. It didn't work, it soon gets replaced by the old image. I have tried adding a different tag to my recently built image and specify to the runner it via `--docker-compose-images-tags`. Didn't work either.

I guess the question is very simple, how can I get a dependency installed inside the `clickhouse/integration-test` image? The answer would be: simply add it to the Dockerfile. Unfortunately, for some reason, it always replaces by this "old" image and doesn't use my recently built one.

tavplubix commented:
Yes, you're right, I just had no ideas what's wrong with docker images in your environment, so I tried to suggest some workarounds (like running the script from outside).

Simply editing the Dockerfile and running `docker build -t image_name` usually works fine for me, but the last time I did it for `clickhouse/integration-test` was a long time ago... I pulled you branch, tried to run the test locally and it did not work this time for some reason. So the issue was not in your environment.

After some debugging I found the reason:
https://github.com/ClickHouse/ClickHouse/blob/4b4c63749afcca4fd8b111618690bb2a0619f581/tests/integration/helpers/cluster.py#L2333-L2343
This code replaces locally built image with pulled one... But it's not really needed for running tests locally (https://github.com/ClickHouse/ClickHouse/pull/33215) and commenting it helps

arthurpassos committed:
fix test errors

arthurpassos commented:
> Yes, you're right, I just had no ideas what's wrong with docker images in your environment, so I tried to suggest some workarounds (like running the script from outside).
> 
> Simply editing the Dockerfile and running `docker build -t image_name` usually works fine for me, but the last time I did it for `clickhouse/integration-test` was a long time ago... I pulled you branch, tried to run the test locally and it did not work this time for some reason. So the issue was not in your environment.
> 
> After some debugging I found the reason:
> 
> https://github.com/ClickHouse/ClickHouse/blob/4b4c63749afcca4fd8b111618690bb2a0619f581/tests/integration/helpers/cluster.py#L2333-L2343
> 
> 
> This code replaces locally built image with pulled one... But it's not really needed for running tests locally (#33215) and commenting it helps

Thank for looking into this. This change + a docker load did the trick. I believe CI will now at least be able to execute the script. I ran it successfully once, but the next runs failed with multiple:
```
pycurl.error: (45, 'bind failed with errno 98: Address already in use')
    crl.perform()
```
I believe this is due to ephemeral port exhaustion, but not sure. Will see how it goes on CI and if needed I'll investigate it tomorrow.

Can you enable CI?

arthurpassos committed:
lower thread and iterations number

arthurpassos committed:
Merge branch 'master' into fix_ares_fd_process_race_condition

arthurpassos commented:
@tavplubix Do you happen to have an opinion about what I shared earlier? Approach in this PR / failure when number of threads is high vs ephemeral ports vs lock

Plus, can you enable CI?

arthurpassos commented:
Btw, why am I still considered a first-time contributor? I already have a couple of PRs merged

tavplubix commented:
> Btw, why am I still considered a first-time contributor?

You are not, GitHub shows "Contributor" label (or I did not get the question)
![Screenshot from 2022-10-20 19-04-15](https://user-images.githubusercontent.com/14847450/197013498-0dfd3806-e6ad-4cbe-9ef4-c29a33247e00.png)

arthurpassos commented:
> > Btw, why am I still considered a first-time contributor?
> 
> You are not, GitHub shows "Contributor" label (or I did not get the question) ![Screenshot from 2022-10-20 19-04-15](https://user-images.githubusercontent.com/14847450/197013498-0dfd3806-e6ad-4cbe-9ef4-c29a33247e00.png)

Whenever I push a new commit, I need to ask someone to enable the CI run. Plus, this message appears:
![image](https://user-images.githubusercontent.com/31894389/197014532-1f603273-3fd5-41d3-b0b5-89172a1ef0c9.png)

tavplubix commented:
For security reasons manual approval is required for all external contributors (not only first-time contributors, I don't know why the message looks like this) if some scripts related to the CI itself are changed (it includes changes in Dockerfiles).

arthurpassos commented:
> For security reasons manual approval is required for all external contributors (not only first-time contributors, I don't know why the message looks like this) if some scripts related to the CI itself are changed (it includes changes in Dockerfiles).

I see, that makes sense. That's why  I was under the impressions this was not always the case

arthurpassos commented:
Something really bad happened in this last CI run lol @tavplubix

tavplubix commented:
Yes, CI was completely broken for a short period of time due to https://github.com/ClickHouse/ClickHouse/pull/42496, it's already reverted, let's restart tests

tavplubix committed:
Merge branch 'master' into fix_ares_fd_process_race_condition

tavplubix commented:
> Do you happen to have an opinion about what I shared earlier? Approach in this PR / failure when number of threads is high vs ephemeral ports vs lock

Although the tests looks good (and they should catch the issue, especially with sanitizers) and server does not crash anymore, I sill don't understand why the solution with `poll` fixes race/crash. In particular, both `ares_fds` and `ares_getsock` call `ares__is_list_empty` intetrnally ([1](https://github.com/c-ares/c-ares/blob/de042a65f492f8a8b0a10bf21494623b9ac345fd/src/lib/ares_getsock.c#L31), [2](https://github.com/c-ares/c-ares/blob/de042a65f492f8a8b0a10bf21494623b9ac345fd/src/lib/ares_fds.c#L30)) and `ares__is_list_empty` does not look thread-safe, because there's `ares__remove_from_list` which can be called from `ares__send_query` which can be called from `next_server` which can be called from `process_timeouts` which can be called from `processfds` which is called from both `ares_process` and `ares_process_fd`. 

So solution with mutex looks much more safe. Maybe it's even possible to unlock mutex before `select`/`poll` and then lock again to avoid long waiting when some requests are being processed and we want to send another one. Also, is it possible to add the unit test (that you posted above) to the CI as well?

arthurpassos commented:
> > Do you happen to have an opinion about what I shared earlier? Approach in this PR / failure when number of threads is high vs ephemeral ports vs lock
> 
> Although the tests looks good (and they should catch the issue, especially with sanitizers) and server does not crash anymore, I sill don't understand why the solution with `poll` fixes race/crash. In particular, both `ares_fds` and `ares_getsock` call `ares__is_list_empty` intetrnally ([1](https://github.com/c-ares/c-ares/blob/de042a65f492f8a8b0a10bf21494623b9ac345fd/src/lib/ares_getsock.c#L31), [2](https://github.com/c-ares/c-ares/blob/de042a65f492f8a8b0a10bf21494623b9ac345fd/src/lib/ares_fds.c#L30)) and `ares__is_list_empty` does not look thread-safe, because there's `ares__remove_from_list` which can be called from `ares__send_query` which can be called from `next_server` which can be called from `process_timeouts` which can be called from `processfds` which is called from both `ares_process` and `ares_process_fd`.
> 
> So solution with mutex looks much more safe. Maybe it's even possible to unlock mutex before `select`/`poll` and then lock again to avoid long waiting when some requests are being processed and we want to send another one. Also, is it possible to add the unit test (that you posted above) to the CI as well?

In regards to `ares__is_list_empty` and `ares__remove_from_list`, those take as parameter an inner structure of the channel. The channel is not shared, we have one per thread. Looking at the impl of those methods, it doesn't seem to touch anything shared.

In any case, I can't explain why it "fixes" either, maybe it's just luck. IIRC, I did a quick benchmark on locking and unlocking twice to avoid waiting and there was not a significant difference, but will test it once more.

I also believe locking is a safer bet, at least until we get a response from `c-ares` folks. I was just trying to avoid unnecessary overhead, but let's go for the safest option.

arthurpassos commented:
> So solution with mutex looks much more safe. Maybe it's even possible to unlock mutex before `select`/`poll` and then lock again to avoid long waiting when some requests are being processed and we want to send another one. Also, is it possible to add the unit test (that you posted above) to the CI as well?

I am not sure that unit test will add much value without a proper DNS server to return the hostnames we need. We could rely on some "well known" domain, but that sounds flaky. Another option I have tried was to query for an empty string and expect no results, but I believe even that returned me a couple of results. Plus, if the query failed, we wouldn't know because.. the expected response is empty anyways

tavplubix commented:
> I am not sure that unit test will add much value without a proper DNS server to return the hostnames we need. We could rely on some "well known" domain, but that sounds flaky.

Yep, you're right. Maybe we could try `1.1.1.1` and `216.58.207.46` (it returns 2 PTR records, however, it may change in future) and do not check actual result, just check that request returns something and it does not crash. But it's probably bad idea to send tons of requests to real DNS from our CI. Or maybe we could add `coredns` to the image that runs unit tests or download `dbms_unit_tests` from an integration test run this specific test. But it's too complicated and is not worth it.

arthurpassos committed:
Make CaresPTRResolver completely thread-safe

arthurpassos commented:
I have done some benchmarking and multiple `mutex.lock` calls (ares_send, ares_getsock and process_fd) yielded a much greater execution time than just a single call enclosing everything. Plus, that would make the code very complex and error prone for future changes.

Not sure you remember, but in the first implementation merged the CaresPTRResolver was a singleton. That caused a very easy to reproduce race condition on the `channel` member. To avoid locking, we then moved to a single instance per thread. Since we couldn't get away with it and ended up adding the lock, I made it a singleton again.

As a side note, I also experimented a decorator pattern implementation to handle the locking and avoid adding yet more complexity into `CaresPTRResolver`. In terms of performance, there was no apparent difference. I didn't pursue that path because some things were unclear and more complex than needed, example: 

1. Should the mutex owned by the decorator be static or not?
2. What if the user decides to make an instance of CaresPTRResolver without the decorator, that would cause us some problems (The provider would not allow that, but still is more complex than needed). 

In any case, just 4 fun, here is the decorator impl:
```
namespace DB {
    class LockedPTRResolver : public DNSPTRResolver
    {
        friend class DNSPTRResolverProvider;
        /*
         * Allow only DNSPTRProvider to instantiate this class
         * */
        struct provider_token {};

    public:
        LockedPTRResolver(std::unique_ptr<DNSPTRResolver> resolver, provider_token);

        std::unordered_set<std::string> resolve(const std::string & ip) override;

        std::unordered_set<std::string> resolve_v6(const std::string & ip) override;

    private:
        // this needs to be owned
        std::unique_ptr<DNSPTRResolver> resolver;

        static std::mutex mutex;
    };
}
```
```
namespace DB {

    std::mutex LockedPTRResolver::mutex;

    LockedPTRResolver::LockedPTRResolver(std::unique_ptr<DNSPTRResolver> resolver_, provider_token)
    : resolver(std::move(resolver_))
    {}

    std::unordered_set<std::string> LockedPTRResolver::resolve(const std::string & ip)
    {
        std::lock_guard guard(mutex);

        return resolver->resolve(ip);
    }

    std::unordered_set<std::string> LockedPTRResolver::resolve_v6(const std::string & ip)
    {
        std::lock_guard guard(mutex);

        return resolver->resolve_v6(ip);
    }

}

```
```
std::shared_ptr<DNSPTRResolver> DNSPTRResolverProvider::get()
{
    static auto resolver = std::make_shared<LockedPTRResolver>(
        std::make_unique<CaresPTRResolver>(
            CaresPTRResolver::provider_token {}
        ),
        LockedPTRResolver::provider_token {}
    );

    return resolver;
}
```

arthurpassos committed:
what

arthurpassos commented:
I messed up with the files comitted, let me fix this before CI run

arthurpassos committed:
removing unwanted files

arthurpassos commented:
Done, can you enable CI? And review :) @tavplubix

arthurpassos commented:
IntegrationTestsRelease1The run was canceled by @clickhouse-ci.
--
IntegrationTestsTsan2The run was canceled by @clickhouse-ci.

[IntegrationTestsRelease1](https://github.com/ClickHouse/ClickHouse/actions/runs/3291357091/jobs/5437360912)
The run was canceled by @clickhouse-ci.
[IntegrationTestsTsan2](https://github.com/ClickHouse/ClickHouse/actions/runs/3291357091/jobs/5437364781)
The run was canceled by @clickhouse-ci.

m1khal3v commented:
Is this fix going to be merged into the 22.8 branch?

arthurpassos commented:
CI looks good. We finally got a response in the [c-ares multi channel/ thread reverse lookup issue](https://github.com/c-ares/c-ares/issues/484). They say c-ares has no notion of threads, but it doesn't have a global state and multithreaded applications should work just fine as long as they don't reference the same channel.

I believe we couldn't find such global state either. So.. it's still a mystery why this happens and why this seems to fix the issue.

Taking a step back, before I suspected it was a race condition, the stack trace was pointing to a corrupted `Poco::Net::IPAddress` object. In a quick investigation, I couldn't find anything weird that could led up to that.

Anyways, the benchmark and python script don't cause ClickHouse to crash anymore, so I guess this should be merged, but with the idea that the error might be somewhere else. Thoughts @tavplubix?

tavplubix reviewed (approved):
OK, let's merge it and see if it fixes the issue

tavplubix commented:
Stateless tests (debug) [3/3] - https://github.com/ClickHouse/ClickHouse/pull/42611
Stress test (asan) - https://github.com/ClickHouse/ClickHouse/issues/41500
Stress test (msan) - https://github.com/ClickHouse/ClickHouse/issues/41831
Stress test (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/42740

den-crane commented:
I would backport it to 22.8. It blocks several big clickhouse customers.

Algunenano commented:
Please beware that this PR broke the build because of the `-Wshorten-64-to-32` warning that was introduced in parallel.

arthurpassos commented:
> Please beware that this PR broke the build because of the `-Wshorten-64-to-32` warning that was introduced in parallel.

I am not sure I follow, what do you mean by "was introduced in parallel"?

Algunenano commented:
> I am not sure I follow, what do you mean by "was introduced in parallel"?

I mean that while this PR was being worked on and reviewed, another PR introduced new warnings in the build. This PR wasn't rebased for a while so it had code that triggered the warnings. The master build is broken as a result.

PhantomPhreak commented:
recent LTS version 22.8.9.24 was build with the fix: https://github.com/ClickHouse/ClickHouse/compare/v22.8.6.71-lts...v22.8.9.24-lts

But is still can reproduce the issue
```
2022.12.01 15:37:12.986472 [ 22681 ] {} <Information> SentryWriter: Sending crash reports is initialized with https://6f33034cfe684dd7a3ab9875e57b1c8d@o388870.ingest.sentry.io/5226277 endpoint and /var/lib/clickhouse/tmp/sentry temp folder (anonymized)
2022.12.01 15:37:13.723989 [ 22681 ] {} <Debug> Application: Skipped file in temporary path /var/lib/clickhouse/tmp/sentry
2022.12.01 15:37:48.069852 [ 22918 ] {} <Fatal> BaseDaemon: ########################################                                                                                                       
2022.12.01 15:37:48.070067 [ 22918 ] {} <Fatal> BaseDaemon: (version 22.8.9.24 (official build), build id: 7813706E2478567D) (from thread 22808) (no query) Received signal Segmentation fault (11)        
2022.12.01 15:37:48.070220 [ 22918 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.                                                                             
2022.12.01 15:37:48.070340 [ 22918 ] {} <Fatal> BaseDaemon: Stack trace: 0x188d1aa3 0xa3df85a 0xa4d8e53 0xa500a27 0xa5000db 0x193a13c6 0x193ae62e 0x193ad627 0x193ac3a1 0xa4ffe23 0xa4ffc3b 0xa4f791c 0xa4fb
248 0xa4fa69c 0xa4f90cc 0x1493467a 0x13ee8658 0x13eeaf36 0x13eebc6c 0xa4b18a8 0xa4b4abd 0x7f1596b0a494 0x7f159684caff                                                                                      
2022.12.01 15:37:48.070554 [ 22918 ] {} <Fatal> BaseDaemon: 2. sallocx @ 0x188d1aa3 in /usr/bin/clickhouse                                                                                                 
2022.12.01 15:37:48.070722 [ 22918 ] {} <Fatal> BaseDaemon: 3. operator delete(void*, unsigned long) @ 0xa3df85a in /usr/bin/clickhouse                                                                    
2022.12.01 15:37:48.070911 [ 22918 ] {} <Fatal> BaseDaemon: 4. std::__1::__hash_table<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_
string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std
::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::__rehash(unsigned long) @ 0xa4d8e53 in /usr/bin/clickhouse                                                         
2022.12.01 15:37:48.071150 [ 22918 ] {} <Fatal> BaseDaemon: 5. std::__1::pair<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<
char> >, void*>*>, bool> std::__1::__hash_table<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<cha
r>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::ch
ar_traits<char>, std::__1::allocator<char> > > >::__emplace_unique_key_args<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::ch
ar_traits<char>, std::__1::allocator<char> > >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::
__1::allocator<char> >&&) @ 0xa500a27 in /usr/bin/clickhouse
2022.12.01 15:37:48.071307 [ 22918 ] {} <Fatal> BaseDaemon: 6. ? @ 0xa5000db in /usr/bin/clickhouse
2022.12.01 15:37:48.071465 [ 22918 ] {} <Fatal> BaseDaemon: 7. ? @ 0x193a13c6 in /usr/bin/clickhouse
2022.12.01 15:37:48.071761 [ 22918 ] {} <Fatal> BaseDaemon: 8. ? @ 0x193ae62e in /usr/bin/clickhouse
2022.12.01 15:37:48.071921 [ 22918 ] {} <Fatal> BaseDaemon: 9. ? @ 0x193ad627 in /usr/bin/clickhouse
2022.12.01 15:37:48.072031 [ 22918 ] {} <Fatal> BaseDaemon: 10. ? @ 0x193ac3a1 in /usr/bin/clickhouse
2022.12.01 15:37:48.072154 [ 22918 ] {} <Fatal> BaseDaemon: 11. DB::CaresPTRResolver::wait() @ 0xa4ffe23 in /usr/bin/clickhouse
2022.12.01 15:37:48.072309 [ 22918 ] {} <Fatal> BaseDaemon: 12. DB::CaresPTRResolver::resolve(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xa4ffc3b in /
usr/bin/clickhouse
2022.12.01 15:37:48.072449 [ 22918 ] {} <Fatal> BaseDaemon: 13. ? @ 0xa4f791c in /usr/bin/clickhouse
2022.12.01 15:37:48.072571 [ 22918 ] {} <Fatal> BaseDaemon: 14. DB::DNSResolver::updateAddress(Poco::Net::IPAddress const&) @ 0xa4fb248 in /usr/bin/clickhouse
2022.12.01 15:37:48.072723 [ 22918 ] {} <Fatal> BaseDaemon: 15. bool DB::DNSResolver::updateCacheImpl<bool (DB::DNSResolver::*)(Poco::Net::IPAddress const&), std::__1::unordered_map<Poco::Net::IPAddress,
unsigned int, std::__1::hash<Poco::Net::IPAddress>, std::__1::equal_to<Poco::Net::IPAddress>, std::__1::allocator<std::__1::pair<Poco::Net::IPAddress const, unsigned int> > >&>(bool (DB::DNSResolver::*&&)
(Poco::Net::IPAddress const&), std::__1::unordered_map<Poco::Net::IPAddress, unsigned int, std::__1::hash<Poco::Net::IPAddress>, std::__1::equal_to<Poco::Net::IPAddress>, std::__1::allocator<std::__1::pai
r<Poco::Net::IPAddress const, unsigned int> > >&, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_tr
aits<char>, std::__1::allocator<char> > const&) @ 0xa4fa69c in /usr/bin/clickhouse
2022.12.01 15:37:48.072835 [ 22918 ] {} <Fatal> BaseDaemon: 16. DB::DNSResolver::updateCache(unsigned int) @ 0xa4f90cc in /usr/bin/clickhouse
2022.12.01 15:37:48.073163 [ 22918 ] {} <Fatal> BaseDaemon: 17. DB::DNSCacheUpdater::run() @ 0x1493467a in /usr/bin/clickhouse
2022.12.01 15:37:48.073319 [ 22918 ] {} <Fatal> BaseDaemon: 18. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x13ee8658 in /usr/bin/clickhouse
2022.12.01 15:37:48.073575 [ 22918 ] {} <Fatal> BaseDaemon: 19. DB::BackgroundSchedulePool::threadFunction() @ 0x13eeaf36 in /usr/bin/clickhouse
2022.12.01 15:37:48.073696 [ 22918 ] {} <Fatal> BaseDaemon: 20. ? @ 0x13eebc6c in /usr/bin/clickhouse
2022.12.01 15:37:48.073840 [ 22918 ] {} <Fatal> BaseDaemon: 21. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa4b18a8 in /usr/bin/clickhouse
2022.12.01 15:37:48.073959 [ 22918 ] {} <Fatal> BaseDaemon: 22. ? @ 0xa4b4abd in /usr/bin/clickhouse
2022.12.01 15:37:48.074136 [ 22918 ] {} <Fatal> BaseDaemon: 23. start_thread @ 0x7494 in /lib/x86_64-linux-gnu/libpthread-2.24.so
2022.12.01 15:37:48.074294 [ 22918 ] {} <Fatal> BaseDaemon: 24. __clone @ 0xe8aff in /lib/x86_64-linux-gnu/libc-2.24.so
2022.12.01 15:37:48.410282 [ 22918 ] {} <Fatal> BaseDaemon: Integrity check of the executable successfully passed (checksum: 2323D2E80DF69B0C447640591C08798B)
2022.12.01 15:37:48.418431 [ 22918 ] {} <Information> SentryWriter: Sending crash report
```

I've created used with following ACL
```
            <networks incl="networks" replace="replace">
                    <host_regexp>^hostname[1-4]-v\.dev\.domain\.xx$</host_regexp>
            </networks>
```

Running a command like following causing a server crash
```
:) select version(),hostname() from clusterAllReplicas('cluster',system,one)
```

```
SELECT
    name,
    value
FROM system.settings
WHERE changed

┌─name──────────────────────────────────────────┬─value──┐
│ use_uncompressed_cache                        │ 0      │
│ distributed_directory_monitor_batch_inserts   │ 1      │
│ load_balancing                                │ random │
│ distributed_aggregation_memory_efficient      │ 1      │
│ max_parallel_replicas                         │ 1      │
│ log_queries                                   │ 1      │
│ readonly                                      │ 0      │
│ max_memory_usage                              │ 0      │
│ use_compact_format_in_distributed_parts_names │ 1      │
│ input_format_skip_unknown_fields              │ 1      │
└───────────────────────────────────────────────┴────────┘

10 rows in set. Elapsed: 0.006 sec. 
```

Internal DNS cache is enabled in the server settings.
I've enabled and triggered a crash report with Sentry for a few times

arthurpassos commented:
@PhantomPhreak I have just tried with a DNS server set up that actually responds to the reverse queries, no crash either.

PhantomPhreak commented:
Hi @arthurpassos 
> Full test command: curl --interface 192.168.0.157 'http://localhost:8123/?query=select+version(),hostname()+from+clusterAllReplicas('test_shard_localhost',system,one)&user=test_dns'

I forgot to mention that i actually used a native client like following:

```
# echo "select version(),hostname() from clusterAllReplicas('statistics',system,one)"| clickhouse-client -u <user> --password <pw> -h <remote host>
```
Posibly it's important for reproducing the issue

>  Also, did it crash right after your select or a couple of mins later?

Without `<networks>` block set, my test command executes instantly. With ACL configured, it hangs for 10-15s, but ClickHouse server keep writing the log and doing some background operations (like pulling data from Kafka), then it's throwing `Application: Child process was terminated by signal 11.` and crashing
